### ISSUE TITLE
TypeError: this.flowConstructor is not a constructor

### ISSUE THREAD

**Issue by cpi-rdesai** [flowjs/ngx-flow#24]
![image](https://user-images.githubusercontent.com/13028263/70392865-1440af00-1a0a-11ea-87bf-3f0d37a4d397.png)

Gettiing this error ,even after injecting provider as shown below

Providers:[{
      provide: FlowInjectionToken,
      useValue: Flow
}]
I have different feature module for an application being created as ng library and loaded as lazy.
i have a common component for upload, which will be used accross lazy feature modules.

**Comment by cpi-rdesai** [comment]
#15 Tried all possible solution from here

**Comment by MartinNuc** [comment]
Where do you provide the FlowInjectionToken? In the core module?

**Comment by cpi-rdesai** [comment]
I have tried several combination
Tried in root module and also in feature modules(lazy loaded) - dint work
Tried giving only in feature module and remove from root module - dint work
... etc

Can we have the forRoot method of dependency injection?, As stated in your previous issue #15 .
And even the typings "@types/flowjs" were not installed , had to separately install it.

**Comment by GitXUser** [comment]
> Where do you provide the FlowInjectionToken? In the core module?

@MartinNuc ,Even I am facing same issue as mentioned by @cpi-rdesai , tried the above combinations mentioned and injecting FlowInjectionToken in core module. It seems to be issue only with AOT build . Do we have a fix or solution for this ?

**Comment by MartinNuc** [comment]
No, I don't have. Imho you should provide flowjs lib through Angular dependency injection container and that's it. Core modules should be perfectly fine for that.

Do you have a repository with a project with this issue so I can take a look? It happens during the build, right? Not during debug, correct?

**Comment by GitXUser** [comment]
> No, I don't have. Imho you should provide flowjs lib through Angular dependency injection container and that's it. Core modules should be perfectly fine for that.
> 
> Do you have a repository with a project with this issue so I can take a look? It happens during the build, right? Not during debug, correct?

@MartinNuc thanks for replying, this happens only when we try to run the deployed AOT build application, not while debugging or even not while running it locally. I'll share a demo version of this issue with you soon, as i'll have to create one considering compliance.

**Comment by MartinNuc** [comment]
Thank you. I have a suspicion about not including the flowjs lib to the bundle after build. Let's see from the example repository. I will try to look into it.

**Comment by cpi-rdesai** [comment]
@MartinNuc @GitXUser Added sample application code here
https://github.com/cpi-rdesai/FlowTest/upload

**Comment by GitXUser** [comment]
> Thank you. I have a suspicion about not including the flowjs lib to the bundle after build. Let's see from the example repository. I will try to look into it.

I have made some changes in @cpi-rdesai  to run and replicate the AOT issue locally without deployment. @MartinNuc  Please have a look at https://github.com/GitXUser/IssueWithAOTBuildFlowJS . The Local build using ng serve works perfectly fine but when we try to run the application with ng serve --prod , error is encountered "this.flowconstructor is not a constructor" , the exact line of error is as below.
![image](https://user-images.githubusercontent.com/36219454/70865011-79eae900-1f7e-11ea-9d56-b4db5e5d02d0.png)

**Comment by GitXUser** [comment]
> > Thank you. I have a suspicion about not including the flowjs lib to the bundle after build. Let's see from the example repository. I will try to look into it.
> 
> I have made some changes in @cpi-rdesai to run and replicate the AOT issue locally without deployment. @MartinNuc Please have a look at https://github.com/GitXUser/AOTIssueWithFlow/ . The Local build using ng serve works perfectly fine but when we try to run the application with ng serve --prod , error is encountered "this.flowconstructor is not a constructor" , the exact line of error is as below.
> ![image](https://user-images.githubusercontent.com/36219454/70865011-79eae900-1f7e-11ea-9d56-b4db5e5d02d0.png)

@MartinNuc  did you get chance to look at this ? Please let us know. We like stuck for more than a week trying to fix this issue .

**Comment by MartinNuc** [comment]
Not yet. Finishing stuff in my job before Christmas 🙈 I expect to look into it next week on like Monday probably.

**Comment by GitXUser** [comment]
> Not yet. Finishing stuff in my job before Christmas 🙈 I expect to look into it next week on like Monday probably.

I found the solution, below are couple of findings while fixing this :
1. I went through the sample application i.e flowjs/ngx-flow repository , it works perfectly fine with and without --prod flag.
2. The Flow lib is injected  App module by using useValue i.e
providers: [    {      provide: FlowInjectionToken,      useValue: Flow    } ]
this works fine with ^8.0.0 which is in this repo sample.
3. BUT dint work for NG 7.X.X version which is in my sample provided.
the fix i could find is **using 'useFactory' instead of 'useValue'**
providers: [    {      provide: FlowInjectionToken,      useFactory: MyFlowFactory    }]

Reasons is most probably : values in metadata must be statically analyzable for AOT.

Reference and helpful links : 
https://stackoverflow.com/questions/46964243/angular-can-injectiontoken-be-used-with-aot-compilation
https://github.com/angular/angular/issues/19154

@cpi-rdesai  Hope this helps you.
@MartinNuc  Thanks for your efforts . Merry Christmas.
